src/main.rs

```diff
@@ -0,0 +1,141 @@
+            Ok(records) => records,
+            Err(e) => {
+                eprintln!("Failed to fetch Consul DNS records: {}", e);
+
+                return;
+            }
+        };
+
+        // Fetch DNS tags from the services in Consul
+        println!("Fetching DNS tags from the services in Consul");
+        let fetched_dns_records = match consul_client.fetch_service_tags(&mut consul_index).await {
+            Ok(tags) => tags,
+            Err(e) => {
+                eprintln!("Failed to fetch Consul DNS tags: {}", e);
+                return;
+            }
+        };
+        println!(
+            "Consul DNS tags fetched successfully. Length: {}",
+            fetched_dns_records.len()
+        );
+
+        let mut all_success = true;
+
+        for fetched_dns_record in &fetched_dns_records {
+            if !dns_state.values().any(|r| r == fetched_dns_record) {
+                // Create the record on the DNS provider
+                let record = match dns_provider
+                    .update_or_create_dns_record(fetched_dns_record)
+                    .await
+                {
+                    Ok(record) => record,
+                    Err(e) => {
+                        eprintln!("Failed to update or create DNS record: {}", e);
+                        all_success = false;
+                        continue;
+                    }
+                };
+
+                // Store the record in Consul
+                match consul_client
+                    .store_dns_record(record.id, fetched_dns_record)
+                    .await
+                {
+                    Ok(_) => println!("DNS record stored in Consul"),
+                    Err(e) => eprintln!("Failed to store DNS record in Consul: {}", e),
+                }
+            }
+        }
+
+        // Delete DNS records from Consul state that are not in the fetched_dns_records
+        for (record_id, record) in dns_state.iter() {
+            if !fetched_dns_records
+                .iter()
+                .any(|fetched_record| fetched_record == record)
+            {
+                // Delete the record from the DNS provider
+                match dns_provider.delete_dns_record(record_id).await {
+                    Ok(_) => (),
+                    Err(e) => {
+                        eprintln!("Failed to delete DNS record: {}", e);
+                        all_success = false;
+                        continue;
+                    }
+                };
+
+                // Delete the record from Consul state
+                match consul_client.delete_dns_record(record_id).await {
+                    Ok(_) => println!("DNS record deleted from Consul"),
+                    Err(e) => eprintln!("Failed to delete DNS record from Consul: {}", e),
+                }
+            }
+        }
+
+        // Release Lock
+        if consul_client.drop_lock().await.is_err() {
+            eprintln!("Failed to release Consul lock");
+        }
```

Comment:
Once we're using the actual Consul locking support there is no reason to drop the lock.  We should simply continue being the instance that does long-polling until we somehow lose the lock.